<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Finding Novel Exons by Princess Kuteyi</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="capstone rough code_files/libs/clipboard/clipboard.min.js"></script>
<script src="capstone rough code_files/libs/quarto-html/quarto.js"></script>
<script src="capstone rough code_files/libs/quarto-html/popper.min.js"></script>
<script src="capstone rough code_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="capstone rough code_files/libs/quarto-html/anchor.min.js"></script>
<link href="capstone rough code_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="capstone rough code_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="capstone rough code_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="capstone rough code_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="capstone rough code_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Finding Novel Exons by Princess Kuteyi</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="finding-novel-exons" class="level2">
<h2 class="anchored" data-anchor-id="finding-novel-exons">Finding Novel Exons</h2>
<p>The beginning portion of this script is to load the packages needed and also loading the data for the analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: stats4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'S4Vectors' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'GenomeInfoDb' was built under R version 4.3.2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rtracklayer)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Repitools)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Repitools'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:rtracklayer':

    blocks</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SummarizedExperiment</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'SummarizedExperiment' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: MatrixGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: matrixStats</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'MatrixGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Biobase</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Biobase'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:MatrixGenerics':

    rowMedians</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:matrixStats':

    anyMissing, rowMedians</code></pre>
</div>
</div>
</section>
<section id="downloading-biocmanager-and-recount3" class="level2">
<h2 class="anchored" data-anchor-id="downloading-biocmanager-and-recount3">Downloading BiocManager and Recount3</h2>
<p>Using Bioconductor and Recount3 is essential for this project so the first step is to download them.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"recount3"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.1 (2023-06-16)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package(s) not installed when version(s) same as or greater than current; use
  `force = TRUE` to re-install: 'recount3'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Old packages: 'Biostrings', 'dbplyr', 'DESeq2', 'digest', 'GenomeInfoDb',
  'GenomicFeatures', 'ggsci', 'globals', 'Hmisc', 'igraph', 'lattice',
  'multcompView', 'pkgbuild', 'processx', 'ragg', 'remotes', 'S4Arrays',
  'sass', 'systemfonts', 'tidyselect', 'tinytex', 'tm', 'zlibbioc'</code></pre>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"snapcount"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.1 (2023-06-16)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package(s) not installed when version(s) same as or greater than current; use
  `force = TRUE` to re-install: 'snapcount'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Old packages: 'Biostrings', 'dbplyr', 'DESeq2', 'digest', 'GenomeInfoDb',
  'GenomicFeatures', 'ggsci', 'globals', 'Hmisc', 'igraph', 'lattice',
  'multcompView', 'pkgbuild', 'processx', 'ragg', 'remotes', 'S4Arrays',
  'sass', 'systemfonts', 'tidyselect', 'tinytex', 'tm', 'zlibbioc'</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"Repitools"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.1 (2023-06-16)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package(s) not installed when version(s) same as or greater than current; use
  `force = TRUE` to re-install: 'Repitools'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Old packages: 'Biostrings', 'dbplyr', 'DESeq2', 'digest', 'GenomeInfoDb',
  'GenomicFeatures', 'ggsci', 'globals', 'Hmisc', 'igraph', 'lattice',
  'multcompView', 'pkgbuild', 'processx', 'ragg', 'remotes', 'S4Arrays',
  'sass', 'systemfonts', 'tidyselect', 'tinytex', 'tm', 'zlibbioc'</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"recount3"</span>)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"snapcount"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Find available projects to utilize. The focus is human samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>human_projects <span class="ot">&lt;-</span> <span class="fu">available_projects</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-03-21 17:26:50.385019 caching file sra.recount_project.MD.gz.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-03-21 17:26:50.681811 caching file gtex.recount_project.MD.gz.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-03-21 17:26:50.935126 caching file tcga.recount_project.MD.gz.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(human_projects)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8742    6</code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(human_projects)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    project organism file_source     project_home project_type n_samples
1 SRP107565    human         sra data_sources/sra data_sources       216
2 SRP149665    human         sra data_sources/sra data_sources         4
3 SRP017465    human         sra data_sources/sra data_sources        23
4 SRP119165    human         sra data_sources/sra data_sources         6
5 SRP133965    human         sra data_sources/sra data_sources        12
6 SRP096765    human         sra data_sources/sra data_sources         7</code></pre>
</div>
</div>
<p>Pick the project and sample to analyze.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset the project of interest</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a> potential_info <span class="ot">&lt;-</span> <span class="fu">subset</span>(</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a> human_projects,</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>   project <span class="sc">==</span> <span class="st">"SRP107565"</span> <span class="sc">&amp;</span> project_type <span class="sc">==</span> <span class="st">"data_sources"</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>potential_info</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    project organism file_source     project_home project_type n_samples
1 SRP107565    human         sra data_sources/sra data_sources       216</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># utilize gtext as the file source</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="fu">subset</span>(human_projects, file_source <span class="sc">==</span> <span class="st">"gtex"</span> <span class="sc">&amp;</span> project_type <span class="sc">==</span> <span class="st">"data_sources"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             project organism file_source      project_home project_type
8678  ADIPOSE_TISSUE    human        gtex data_sources/gtex data_sources
8679          MUSCLE    human        gtex data_sources/gtex data_sources
8680    BLOOD_VESSEL    human        gtex data_sources/gtex data_sources
8681           HEART    human        gtex data_sources/gtex data_sources
8682           OVARY    human        gtex data_sources/gtex data_sources
8683          UTERUS    human        gtex data_sources/gtex data_sources
8684          VAGINA    human        gtex data_sources/gtex data_sources
8685          BREAST    human        gtex data_sources/gtex data_sources
8686            SKIN    human        gtex data_sources/gtex data_sources
8687  SALIVARY_GLAND    human        gtex data_sources/gtex data_sources
8688           BRAIN    human        gtex data_sources/gtex data_sources
8689   ADRENAL_GLAND    human        gtex data_sources/gtex data_sources
8690         THYROID    human        gtex data_sources/gtex data_sources
8691            LUNG    human        gtex data_sources/gtex data_sources
8692          SPLEEN    human        gtex data_sources/gtex data_sources
8693        PANCREAS    human        gtex data_sources/gtex data_sources
8694       ESOPHAGUS    human        gtex data_sources/gtex data_sources
8695         STOMACH    human        gtex data_sources/gtex data_sources
8696           COLON    human        gtex data_sources/gtex data_sources
8697 SMALL_INTESTINE    human        gtex data_sources/gtex data_sources
8698        PROSTATE    human        gtex data_sources/gtex data_sources
8699          TESTIS    human        gtex data_sources/gtex data_sources
8700           NERVE    human        gtex data_sources/gtex data_sources
8701       PITUITARY    human        gtex data_sources/gtex data_sources
8702           BLOOD    human        gtex data_sources/gtex data_sources
8703           LIVER    human        gtex data_sources/gtex data_sources
8704          KIDNEY    human        gtex data_sources/gtex data_sources
8705    CERVIX_UTERI    human        gtex data_sources/gtex data_sources
8706  FALLOPIAN_TUBE    human        gtex data_sources/gtex data_sources
8707         BLADDER    human        gtex data_sources/gtex data_sources
8708        STUDY_NA    human        gtex data_sources/gtex data_sources
8709     BONE_MARROW    human        gtex data_sources/gtex data_sources
     n_samples
8678      1293
8679       881
8680      1398
8681       942
8682       195
8683       159
8684       173
8685       482
8686      1940
8687       178
8688      2931
8689       274
8690       706
8691       655
8692       255
8693       360
8694      1577
8695       384
8696       822
8697       193
8698       263
8699       410
8700       659
8701       301
8702      1048
8703       251
8704        98
8705        19
8706         9
8707        21
8708       133
8709       204</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>rse_gene_SRP107565 <span class="ot">&lt;-</span> <span class="fu">create_rse</span>(potential_info)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-03-21 17:26:52.865628 downloading and reading the metadata.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-03-21 17:26:53.092224 caching file sra.sra.SRP107565.MD.gz.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-03-21 17:26:53.357173 caching file sra.recount_project.SRP107565.MD.gz.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-03-21 17:26:53.652025 caching file sra.recount_qc.SRP107565.MD.gz.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-03-21 17:26:53.916591 caching file sra.recount_seq_qc.SRP107565.MD.gz.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The 'url'
&lt;http://duffel.rail.bio/recount3/human/data_sources/sra/metadata/65/SRP107565/sra.recount_pred.SRP107565.MD.gz&gt;
does not exist or is not available.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-03-21 17:26:54.443405 downloading and reading the feature information.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-03-21 17:26:54.674833 caching file human.gene_sums.G026.gtf.gz.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-03-21 17:26:54.985671 downloading and reading the counts: 216 samples across 63856 features.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-03-21 17:26:55.203191 caching file sra.gene_sums.SRP107565.G026.gz.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-03-21 17:26:57.014823 constructing the RangedSummarizedExperiment (rse) object.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>rse_gene_SRP107565</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 63856 216 
metadata(8): time_created recount3_version ... annotation recount3_url
assays(1): raw_counts
rownames(63856): ENSG00000278704.1 ENSG00000277400.1 ...
  ENSG00000182484.15_PAR_Y ENSG00000227159.8_PAR_Y
rowData names(10): source type ... havana_gene tag
colnames(216): SRR5579425 SRR5579426 ... SRR5579421 SRR5579422
colData names(168): rail_id external_id ... recount_seq_qc.errq
  BigWigURL</code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">metadata</span>(rse_gene_SRP107565)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$time_created
[1] "2024-03-21 17:26:57 EDT"

$recount3_version
          package ondiskversion loadedversion
recount3 recount3        1.12.0        1.12.0
                                                                                  path
recount3 /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library/recount3
                                                                            loadedpath
recount3 /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library/recount3
         attached is_base       date       source md5ok
recount3     TRUE   FALSE 2023-10-26 Bioconductor    NA
                                                                      library
recount3 /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

$project
[1] "SRP107565"

$project_home
[1] "data_sources/sra"

$type
[1] "gene"

$organism
[1] "human"

$annotation
[1] "gencode_v26"

$recount3_url
[1] "http://duffel.rail.bio/recount3"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(rse_gene_SRP107565)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 63856   216</code></pre>
</div>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowRanges</span>(rse_gene_SRP107565)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 63856 ranges and 10 metadata columns:
                             seqnames            ranges strand |   source
                                &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;
         ENSG00000278704.1 GL000009.2       56140-58376      - |  ENSEMBL
         ENSG00000277400.1 GL000194.1      53590-115018      - |  ENSEMBL
         ENSG00000274847.1 GL000194.1      53594-115055      - |  ENSEMBL
         ENSG00000277428.1 GL000195.1       37434-37534      - |  ENSEMBL
         ENSG00000276256.1 GL000195.1       42939-49164      - |  ENSEMBL
                       ...        ...               ...    ... .      ...
  ENSG00000124334.17_PAR_Y       chrY 57184101-57197337      + |   HAVANA
  ENSG00000185203.12_PAR_Y       chrY 57201143-57203357      - |   HAVANA
   ENSG00000270726.6_PAR_Y       chrY 57190738-57208756      + |   HAVANA
  ENSG00000182484.15_PAR_Y       chrY 57207346-57212230      + |   HAVANA
   ENSG00000227159.8_PAR_Y       chrY 57212184-57214397      - |   HAVANA
                               type bp_length     phase                gene_id
                           &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;            &lt;character&gt;
         ENSG00000278704.1     gene      2237      &lt;NA&gt;      ENSG00000278704.1
         ENSG00000277400.1     gene      2179      &lt;NA&gt;      ENSG00000277400.1
         ENSG00000274847.1     gene      1599      &lt;NA&gt;      ENSG00000274847.1
         ENSG00000277428.1     gene       101      &lt;NA&gt;      ENSG00000277428.1
         ENSG00000276256.1     gene      2195      &lt;NA&gt;      ENSG00000276256.1
                       ...      ...       ...       ...                    ...
  ENSG00000124334.17_PAR_Y     gene      2504      &lt;NA&gt; ENSG00000124334.17_P..
  ENSG00000185203.12_PAR_Y     gene      1054      &lt;NA&gt; ENSG00000185203.12_P..
   ENSG00000270726.6_PAR_Y     gene       773      &lt;NA&gt; ENSG00000270726.6_PA..
  ENSG00000182484.15_PAR_Y     gene      4618      &lt;NA&gt; ENSG00000182484.15_P..
   ENSG00000227159.8_PAR_Y     gene      1306      &lt;NA&gt; ENSG00000227159.8_PA..
                                        gene_type   gene_name       level
                                      &lt;character&gt; &lt;character&gt; &lt;character&gt;
         ENSG00000278704.1         protein_coding  BX004987.1           3
         ENSG00000277400.1         protein_coding  AC145212.2           3
         ENSG00000274847.1         protein_coding  AC145212.1           3
         ENSG00000277428.1               misc_RNA       Y_RNA           3
         ENSG00000276256.1         protein_coding  AC011043.1           3
                       ...                    ...         ...         ...
  ENSG00000124334.17_PAR_Y         protein_coding        IL9R           2
  ENSG00000185203.12_PAR_Y              antisense      WASIR1           2
   ENSG00000270726.6_PAR_Y   processed_transcript AJ271736.10           2
  ENSG00000182484.15_PAR_Y transcribed_unproces..      WASH6P           2
   ENSG00000227159.8_PAR_Y unprocessed_pseudogene    DDX11L16           2
                                    havana_gene         tag
                                    &lt;character&gt; &lt;character&gt;
         ENSG00000278704.1                 &lt;NA&gt;        &lt;NA&gt;
         ENSG00000277400.1                 &lt;NA&gt;        &lt;NA&gt;
         ENSG00000274847.1                 &lt;NA&gt;        &lt;NA&gt;
         ENSG00000277428.1                 &lt;NA&gt;        &lt;NA&gt;
         ENSG00000276256.1                 &lt;NA&gt;        &lt;NA&gt;
                       ...                  ...         ...
  ENSG00000124334.17_PAR_Y OTTHUMG00000022720.1         PAR
  ENSG00000185203.12_PAR_Y OTTHUMG00000022676.3         PAR
   ENSG00000270726.6_PAR_Y OTTHUMG00000184987.2         PAR
  ENSG00000182484.15_PAR_Y OTTHUMG00000022677.5         PAR
   ENSG00000227159.8_PAR_Y OTTHUMG00000022678.1         PAR
  -------
  seqinfo: 374 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>